src/player.js

```diff
@@ -1286,6 +1296,25 @@ export default class Player extends FakeEventTarget {
     return false;
   }
 
+  /**
+   * update the ABR restriction config
+   * @function _applyABRRestriction
+   * @returns {void}
+   * @param {Object} config - new config which configure for checking if it relevant config has changed
+   * @private
+   */
+  _applyABRRestriction(config: Object): void {
+    if (Utils.Object.hasPropertyPath(config, 'abr.restrictions') && this._allTracks) {
+      const videoTracks = this._allTracks.filter(track => track instanceof VideoTrack);
+      const newVideoTracks = this._filterVideoTrackNotInRange(videoTracks);
+      const currentVideoTracks = this._availableTracks.filter(track => track instanceof VideoTrack);
+      this._engine.applyABRRestriction(this._config.abr);
+      if (JSON.stringify(currentVideoTracks) !== JSON.stringify(newVideoTracks)) {
+        this._availableTracks = this._allTracks.filter(track => !(track instanceof VideoTrack)).concat(newVideoTracks);
+        this.dispatchEvent(new FakeEvent(CustomEventType.TRACKS_CHANGED, {tracks: newVideoTracks}));
```

Comment:
Isn't it is better to do this actual filtering in the adapter itself, and get notified about the new tracks.
The engine should know better what tracks are available to it after applying filtering.
I think Shaka does this, while maybe hlsjs doesn't. Maybe good chance to donate this to hlsjs :-)

Reply:
the bitrate is not the only filterable attribute, we will add also height and width in the future and `_filterVideoTrackNotInRange` will become increasingly more complex.

Reply:
I don't think there is reason to use it on engine level if it's relevant for all engines, adapters return the levels from manifest the restriction relevant for their adaptive selection

Reply:
Tracks still available if you want to select a track, not from the range - it's possible in Shaka

Reply:
What I mean is that we assume the internal filtering of each engine will yield same result as this filtering does, and this can lead to incorrect data.